Implement ability to run external work

wddgit opened the pull request:
Implements Framework level support for modules that run in new ways. For example, they might be used to run GPUs, FPGAs, or GEANT V. Initially we expect these modules to be used in prototypes and tests that will direct the path of future development.

These new modules are declared as EDFilters or EDProducers in the global or stream namespace with a template parameter of type ExternalWork. These special modules will have a new acquire method in which they are expected to get products from the event and pass that data to an external worker. The external worker is also passed a WaitingTaskWithArenaHolder that it uses to signal the Framework when it is done and also to pass exceptions back to the Framework.  After the external work is complete, the produce or filter method is called as before. A critically important point is that the acquire method returns and is removed from the stack without waiting for the external work to complete. 

Except for the unit tests included in this PR, no modules currently use this new External Work feature so it should not have any effect on any pre-existing tests.  

(As a separate PR that will follow this one, I am working on the stall grapher code so that stall grapher will properly account for the new signals related to the acquire method)

wddgit committed:
Implement ability to run external work

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2388

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2388/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2388/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

wddgit committed:
Fix code-checks issue

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2389

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

FWCore/Concurrency
FWCore/Framework
FWCore/Integration
FWCore/MessageService
FWCore/ServiceRegistry
FWCore/Services


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald this is something you requested to watch as well.
@davidlange6, @slava77 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

wddgit commented:
@fwyzard Mostly this is the same as we presented at the Core meeting. In response to your comments, we did change the interface so the Framework gets the arena automatically and the module writer does not need to worry about that. Definitely an improvement.

fwyzard commented:
thanks @wddgit .

@rovere @felicepantaleo @vkhristenko this interests you as well.

fwyzard commented on FWCore/Framework/interface/global/implementors.h:
```diff
@@ -53,7 +55,9 @@ namespace edm {
           caches_.resize(iNStreams,static_cast<C*>(nullptr));
+          preallocStreamsExtend(iNStreams);
         }
+        virtual void preallocStreamsExtend(unsigned int iNStreams) {}
```
what is the new `preallocStreamsExtend` method supposed to be used for ?

fwyzard replied:
Then I would suggest to drop `Extend` from the name, since this will likely be called only once at the beginning of the job, and I would favour a name like `preallocate` or `reserve` to match what we use for services or for vectors.

Is this only for `global` and `limited`, or also for `stream` and `one` modules ?
For a `limited` module, does it return the number of streams or the maximum concurrency of the module ?

Finally, this seems to give an alternative to the `StreamCache` approach: one can just have a `vector` as a data member, `resize()` it in the `preallocate` method, and use it with the `StreamID` in the `analyze`/`produce`/`filter` method, right ?

wddgit replied:
I will change the name to preallocate. That is much better. (The reason we used extend was related to things internal to the Framework that a module writer should not need to know about ...). It is only for global and limited modules. The other kinds of modules should not need to know the number of streams. For both limited and global, it refers to the total number of streams not the maximum concurrency. Your third point is correct. If you look inside of StreamCacheHolder it just contains a vector of pointers to the cached objects in addition to some other things (mainly a larger interface of available functions). It is supposed to make it easier for a module writer not to have to implement the code related to the vector. I used preallocate in the test code because I wanted to construct a single object (the WaitingServer which handles all streams) and also wanted it initialized earlier than the point where the StreamCache is available.

fwyzard commented:
Unrelated question: is there a way I can be notified whenever the framework implements a new signal, so that I can check if the `FastTimerService` needs to support it ?

vkhristenko commented on FWCore/Framework/src/Worker.h:
```diff
@@ -441,7 +463,95 @@ namespace edm {
+            auto& es = m_es;
+            auto parentContext = m_parentContext;
+            auto serviceToken = m_serviceToken;
+            queue.push( [worker, &principal, &es, parentContext, serviceToken, holder = m_holder]()
```
If I may try to answer:

After execute is finished, the AcquireTask will be destroyed (as tbb::task is destroyed after execute has completed and there is no recycling).

So, you can not capture `this` directly as the local stack of the `execute` function will be gone but the module which is called is supposed to have access to some of these variables after execute is finished.

Vk

fwyzard replied:
Ah, I see, you cannot capture individual member variables, unless you assign them.
This would work:
```
[worker=m_worker, &principal=m_principal, &es=m_es, parentContext=m_parentContext, serviceToken=m_serviceToken, holder = m_holder](){...}
```
or even this if one wants to be confusing:
```
[m_worker=m_worker, &m_principal=m_principal, &m_es=m_es, m_parentContext=m_parentContext, m_serviceToken=m_serviceToken, m_holder = m_holder](){...}
```
but this does not:
```
[m_worker, &m_principal, &m_es, m_parentContext, m_serviceToken, m_holder](){...}

Thanks,
.A
```

wddgit replied:
I agree with what was said above. This happened because I copy pasted this code from code written prior to C++14, so it originally had to use the C++11 syntax, then I added the last variable using the C++14 syntax. It would be better to be consistent and use the C++14 syntax throughout. I will change this in both places.

cmsbuild commented:
-1

Tested at: 1bb2a63879537a7c4d01c7d8e490e9ddfa3b0089


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/9b389656fc7abd70a8507259b1eb6ee57a00d319
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24901/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24901/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24901/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-30-2300/src/FWCore/Framework/test/stubs/TestLimitedFilters.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-30-2300/src/FWCore/Framework/test/stubs/TestFilterModule.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-30-2300/src/FWCore/Framework/test/stubs/LoadableDummyESSource.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-30-2300/src/FWCore/Framework/test/stubs/TestESDummyDataAnalyzer.cc 
In file included from /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-30-2300/src/FWCore/Framework/test/stubs/ToyModules.cc:14:
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-30-2300/src/FWCore/Framework/interface/stream/EDProducer.h:81:31: error: argument type 'edm::WaitingTaskWithArenaHolder' is incomplete
        base->acquire(ev, es, holder);
                              ^~~~~~
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-11-30-2300/src/FWCore/Framework/interface/stream/implementors.h:33:9: note: forward declaration of 'edm::WaitingTaskWithArenaHolder'
  class WaitingTaskWithArenaHolder;
        ^
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/9b389656fc7abd70a8507259b1eb6ee57a00d319
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24901/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24901/git-merge-result

cmsbuild commented:
Comparison job queued.

wddgit commented:
@fwyzard  To be notified of all new signals, you can sign up to get notifications for PRs modifying the package FWCore/ServiceRegistry and look for changes in the file ActivityRegistry.h. I looked and there were about a half dozen PRs affecting ServiceRegistry per year during the last few years. The link below tells how to do that. I will also try to remember to direct a comment to you highlighting changes that add signals (but that is not as reliable ...). We don't often add new signals.

http://cms-sw.github.io/faq.html#how-do-i-receive-notifications-about-pull-requests-for-a-given-package--subsystem

Dr15Jones commented on FWCore/Framework/src/Worker.cc:
```diff
@@ -377,4 +378,49 @@ namespace edm {
+    try {
+      convertException::wrap([&]()
+      {
+        this->implDoAcquire(ep, es, &moduleCallingContext_, holder);
```
Should just before this line you call `ranAcquireWithoutException_= false`?

wddgit replied:
This was being set to false in reset function so it gets reset every event, even the case of dealing with exceptions from prefetching.

    void reset() {
      cached_exception_ = std::exception_ptr();
      state_ = Ready;
      waitingTasks_.reset();
      workStarted_ = false;
      ranAcquireWithoutException_ = false;
      numberOfPathsLeftToRun_ = numberOfPathsOn_;
    }

wddgit replied:
ranAcquireWithoutException_ is now set in different places after the commit which wraps a task to handle the externalWork exceptions.

fwyzard commented:
Thanks, I've done that.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -441,7 +463,95 @@ namespace edm {
+          }
+        }
+
+        m_worker->runAcquireAfterAsyncPrefetch(excptr,
```
`runAcquireAfterAsyncPrefetch` first checks `excptr` and will not call `acquire(...)` if it is set. It also handles all the work needed to propagate the exception if it is set.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -723,13 +833,21 @@ namespace edm {
+            this, ep, es, streamID, parentContext, context));
+          moduleTask = new (tbb::task::allocate_root()) AcquireTask<T>(
+            this, ep, es, parentContext, std::move(runTaskHolder));
+        }
```
Could this logic be simplified to
```C++
auto moduleTask = new (tbb::task::allocate_root()) RunModuleTask<T>(
                            this, ep,es,streamID,parentContext,context);
if (T::isEvent_ && hasAcquire()) {
 moduleTask = new (tbb::task::allocate_root()) AcquireTask<T>(
             this, ep, es, parentContext, edm::WaitingTaskWithArenaHolder{moduleTask});
}
prefetchAsync(moduleTask, parentContext, ep);
```

wddgit replied:
Yes and I did it, but the same code was also affected by the changed way of handling external work exceptions.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -742,7 +860,22 @@ namespace edm {
       TransitionIDValue<typename T::MyPrincipal> idValue(ep);
       if(shouldRethrowException(*iEPtr, parentContext, T::isEvent_, idValue)) {
-        exceptionPtr = *iEPtr;
+        if (ranAcquireWithoutException_) {
```
Any chance this could be done in the task wrapper? Else, maybe move to a different function to keep this template function from growing.

wddgit replied:
I created a task wrapper to handle this and pushed a commit with that change. I could not figure out a way to get rid of the bool, but succeeded  in removing all changes from runAcquireAfterAsyncPrefetch. This part of the code is no longer modified.

Dr15Jones commented on FWCore/Framework/src/WorkerT.h:
```diff
@@ -83,8 +84,14 @@ namespace edm {
+
+    void implDoAcquire(EventPrincipal const& ep, EventSetup const& c,
+                       ModuleCallingContext const* mcc,
+                       WaitingTaskWithArenaHolder& holder) override;
```
Could this be marked `final`?

wddgit replied:
Yes. I will make that change.

fwyzard commented on FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h:
```diff
@@ -0,0 +1,118 @@
+
+    // ---------- member data --------------------------------
+    WaitingTask* m_task;
+    std::shared_ptr<tbb::task_arena> m_arena;
```
From https://software.intel.com/en-us/node/506359 it doesn't seem that `tbb:task_arena` should be copied around.

wddgit replied:
Below is what I see in the tbb header file for task_arena. Although there is a copy constructor, it does not copy the pointer to the internal arena. Below is the comment in the header and the tbb source implementing the copy constructor looked consistent with that comment.

task_arena(const task_arena &s) // copy settings but not the reference or instance

I do not see an assignment operator or default constructor  for task_arena either. I think it really needs to be a shared_ptr.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24901/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2715851
 - DQMHistoTests: Total failures: 21
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2715654
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.07000000002 KiB( 23 files compared)
 - Checked 111 log files, 8 edm output root files, 27 DQM output files

wddgit committed:
Fix CLANG Build error

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2402

cmsbuild commented:
Pull request #21624 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

wddgit commented:
please test

there are more changes coming in response to comments, but I want to see if this will now get past all the automatic tests.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f8d8aba5eac99703ebfabd5da9755ddc3020fbb1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24915/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/bb52437ebfcd0537a73be3feed966b561f5a6709
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24915/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24915/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24915/summary.html

There are some workflows for which there are errors in the baseline:
 10824.0 step 5
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 26
 - DQMHistoTests: Total histograms compared: 2606833
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2606661
 - DQMHistoTests: Total skipped: 171
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.02000000002 KiB( 22 files compared)
 - Checked 106 log files, 8 edm output root files, 26 DQM output files

wddgit commented:
hold

while I put in several commits in response to comments

Dr15Jones commented:
hold
Since @wddgit request for hold didn't seem to work

cmsbuild commented:
Pull request has been put on hold by  @Dr15Jones
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

wddgit committed:
Use C++14 Lambda capture consistently

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2416

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2416/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2416/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

wddgit committed:
Rename preallocStreamsExtend to preallocate

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2421

cmsbuild commented:
Pull request #21624 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

wddgit commented on FWCore/Framework/src/Event.cc:
```diff
@@ -69,6 +71,14 @@ namespace edm {
+      }
+    } else if (gotBranchIDsFromAcquire) {
+      gotBranchIDsFromAcquire_ = gotBranchIDsFromAcquire;
+      gotBranchIDsFromAcquire_->clear();
```
We get to this point (line 80) just before the acquire method is called. The module actually holds the gotBranchIDsFromAcquire container and this gives the Event a pointer to it. The container has to be in the module because the Event object will go away after acquire is done, but the contents of the container need to be remembered until produce it called (with a different Event object). At this point, it is before anything has been done on the event and the container would still contain entries from the previous event this stream processed, so it must be cleared. I will try adding comments here to clarify the logic.

cmsbuild commented:
+1
Tested at: a08ff238c992ee142e6df2c742219a12d8508458
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24935/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/632b6c042baf87d350cb619157835bedf44e4be0
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24935/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24935/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24935/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21624/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835085
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2834906
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.6100000001 KiB( 22 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

wddgit committed:
Add comment in setProducer

wddgit committed:
Declare implDoAcquire as final

wddgit committed:
Move code from header to source to keep task_arena.h out of Worker.h

wddgit committed:
Wrap external work exception handling in task

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2473

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2473/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2473/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

wddgit committed:
Make code-checks happy

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2474

cmsbuild commented:
Pull request #21624 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

wddgit commented:
please test
@Dr15Jones can you unhold this please? I think I put in changes or responses for all the comments and this is ready to move forward. The first commit is the original pull request and the subsequent commits are all in response to the comments. The only one where I was unsure whether I did what you wanted was the one that added the wrapper task to handle the external work exceptions.

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented:
unhold

cmsbuild commented:
-1

Tested at: 894b0a55b0a76b5db315c047c92c273737c92c4e


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/0179dc54fdff21005908765fa22a5a152179e983
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24978/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24978/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24978/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-06-1100/src/FWCore/Framework/test/TestDriver.cpp 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-06-1100/src/FWCore/Framework/test/TestDriver.cpp 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-06-1100/src/FWCore/Framework/test/TestDriver.cpp 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-06-1100/src/FWCore/Framework/test/TestDriver.cpp 
In file included from /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-06-1100/src/FWCore/Concurrency/src/WaitingTaskWithArenaHolder.cc:11:
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-06-1100/src/FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h:26:3: error: declaration conflicts with target of using declaration already in scope
  class task_arena;
  ^
/cvmfs/cms-ib.cern.ch/nweek-02501/slc6_amd64_gcc630/external/tbb/2018_U1/include/tbb/task_arena.h:173:7: note: target of using declaration
class task_arena : public internal::task_arena_base {
      ^
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/0179dc54fdff21005908765fa22a5a152179e983
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24978/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24978/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24978/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21624/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835085
 - DQMHistoTests: Total failures: 21
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2834886
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.14000000003 KiB( 23 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

wddgit committed:
Replace forward declaration of task_arena with include of header

TBB is written such that it is hard to forward declare task_arena
because of namespace and using tricks.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2497

cmsbuild commented:
Pull request #21624 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d7a1f7efde59fee04aebea1ea756a0d5dc568e45
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24997/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/2699c4c446b0a1dc00b007abba3ee486763ccc27
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24997/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24997/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/24997/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.7611, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21624/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835085
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2834906
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.79999999985 KiB( 23 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

Dr15Jones commented on FWCore/Framework/src/global/EDAnalyzerBase.cc:
```diff
@@ -179,6 +180,7 @@ namespace edm {
     }
     
     void EDAnalyzerBase::preallocStreams(unsigned int) {}
+    void EDAnalyzerBase::preallocate(unsigned int) {}
```
Let's change these to take the full `PreallocationConfiguration` object. Otherwise, how would someone know what the `unsigned int` represents.

wddgit replied:
Done, I changed the argument to preallocate and preallocStreams.

Dr15Jones commented on FWCore/Framework/src/Event.cc:
```diff
@@ -52,7 +52,9 @@ namespace edm {
-  Event::setProducer( ProducerBase const* iProd, std::vector<BranchID>* previousParentage) {
+  Event::setProducer(ProducerBase const* iProd,
+                     std::vector<BranchID>* previousParentage,
+                     std::vector<BranchID>* gotBranchIDsFromAcquire) {
```
Let's break this into two functions, the original `setProducer` and a new `setProducerForAcquire`. Essentially the `if(previousParentage)` is already trying to make this one method do the function of two methods.

wddgit replied:
Done, I split setProducer into 2 functions.

wddgit committed:
Change argument type of preallocate and preallocStreams

wddgit committed:
Split setProducer function into setProducer and setProducerForAcquire

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2518

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21624 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

wddgit committed:
revert recent change to preallocStreams

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2519

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21624 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+1
Tested at: 17989784f898b6227b477470e22accc568b2c0f0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25018/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/623a8313d815d5fa3415d9cc9fa1f12a66fa6c47
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25018/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25018/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25018/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21624/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835241
 - DQMHistoTests: Total failures: 21
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2835042
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.569999999905 KiB( 23 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

wddgit committed:
Minor simplification in stream module base classes

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/PR-21624/2536

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #21624 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar (and backports should be raised in the release meeting by the corresponding L2)

cmsbuild commented:
+1
Tested at: a2869738d58860d3c1d3ff937086729bb0c4a532
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25035/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/ef598b6d86051da3187d1fe68fdec21f800d5d40
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25035/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25035/git-merge-result

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-21624/25035/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-21624/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2835241
 - DQMHistoTests: Total failures: 21
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2835042
 - DQMHistoTests: Total skipped: 178
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.03999999983 KiB( 23 files compared)
 - Checked 113 log files, 9 edm output root files, 27 DQM output files

davidlange6 commented:
+1
